core/src/main/java/com/yahoo/oak/FirstLevelHashArray.java

```diff
@@ -17,56 +17,76 @@
-                msbForFirstLevelHash, Integer.SIZE); // the second (MSB) will be msbForFirstLevelHash
+                msbForFirstLevelHash, // the second (MSB) will be msbForFirstLevelHash
+                1, // and disregard the first bit as it is the sign bit and always zero
+                Integer.SIZE);
 
         // the size of the hash array is defined by number of MSBs to be used from key hash
         int arraySize = (int) (Math.pow(2, msbForFirstLevelHash));
         this.chunks = new AtomicReferenceArray<>(arraySize);
         this.msbForFirstLevelHash = msbForFirstLevelHash;
 
+        int currentSameRefer = multipleReferenceNum;
+        HashChunk<K, V> c = null;
+        UnionCodec hashIndexCodecForChunk =
+            (lsbForSecondLevelHash == InternalOakHash.USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? this.hashIndexCodec :
+                new UnionCodec(lsbForSecondLevelHash, // the size of the first, as these are LSBs
+                    UnionCodec.AUTO_CALCULATE_BIT_SIZE, Integer.SIZE); // the second (MSB) will be auto-calculated
+        int chunkSize = calculateChunkSize(lsbForSecondLevelHash);
         // initiate chunks
-        for (int i = 0; i < msbForFirstLevelHash; i++) {
-            HashChunk<K, V> c =
-                new HashChunk<>(calculateChunkSize(), externalSize, vMM, kMM, comparator,
-                    keySerializer, valueSerializer, hashIndexCodec);
+        System.out.println("Going to allocate " + chunks.length() + " chunks, of size " + chunkSize);
+        for (int i = 0; i < chunks.length(); i++) {
+            if (currentSameRefer == multipleReferenceNum) {
+                c = new HashChunk<>(chunkSize, externalSize, vMM, kMM, comparator,
+                    keySerializer, valueSerializer, hashIndexCodecForChunk);
+            }
+            System.out.print(i + " ");
             this.chunks.lazySet(i, c);
+            currentSameRefer--;
+            if (currentSameRefer == 0) {
+                currentSameRefer = multipleReferenceNum;
+            }
         }
-
     }
 
-    private int calculateChunkSize() {
+    private int calculateChunkSize(int inputLsbForSecondLevel) {
         // least significant bits remaining
-        int lsbForSecondLevel = Integer.SIZE - msbForFirstLevelHash;
+        int lsbForSecondLevel =
+            (inputLsbForSecondLevel == InternalOakHash.USE_DEFAULT_FIRST_TO_SECOND_BITS_PARTITION)
+                ? Integer.SIZE - msbForFirstLevelHash : inputLsbForSecondLevel;
         // size of HashChunk should be 2^lsbForSecondLevel
         return (int) (Math.pow(2, lsbForSecondLevel) + 1); // +1 to round up
     }
 
     private int calculateKeyHash(K key, ThreadContext ctx) {
-        int hashKey = key.hashCode(); // hash can be positive, zero or negative
-        ctx.operationKeyHash = Math.abs(hashKey); // EntryHashSet doesn't accept negative hashes
         return ctx.operationKeyHash;
     }
```

Comment:
It looks weird that `calculateKeyHash()` doesn't actually calculate anything.
I suggest moving this to `ThreadContext` such that `ThreadContext.calculateKeyHash()` will cache the result internally, and will return the cached result if called again.

Reply:
It is currently done the way as you suggest but in `InternalOakHash` because of the big amount of supporting fields required to calculate the hash (serializers, memory manager, etc.)